packages/git-extractor/src/routes/github/pull/download.ts

```diff
@@ -90,5 +83,34 @@ export async function downloadRepository(
     })
   );
 
+
+  const requestsToMake = pathArray.length
+
+  /**
+   * Check if there is enough of our CodeSandbox Github token rate limit left to be able to
+   * request all the files we need to. If there isn't, then we shouldn't make the Promise.all
+   * request because when the first 403 rate limit comes through, it rejects everything, and
+   * it wastes even more rate limit tries.
+   */
+  const canRequest = await checkRemainingRateLimit(requestsToMake);
+  if (!canRequest) {
+    throw new Error("Can't make axios requests, not enough rate limit remaining")
+  }
+
+  // Then we can request the SHAs of binary files if there is enough rate limit left.
+  await Promise.all(pathArray.map(async (relativePath) => {
```

Comment:
This used to be one big `Promise.all` but it meant we couldn't count the number of binary files + requests needed to be made beforehand, so I've separated it into two - first for processing text files, then when we are ready to fetch the commit shas of the binary files.

Reply:
Nice! 😄 
Sounds like a good call, processing the text files first to make sure we're not wasting our limited calls on binary files!

Reply:
This is very nicely done!!